### DISCUSSION TITLE
I keep stuggeling with with models and builders

### DISCUSSION THREAD

**Discussion by mdietger** [larastan/larastan#1239] [General]
Hi all

So I have an application and I want to add Larastan to it. I only have limited knowledge about static analysis but I get the general idea/concept and really see the added value in a project. Right now I have a Laravel 9 application with Larastan 2 and upon the first run (lvl 1) without adding any docblocks I encountered the following errors.

```
 ------ ----------------------------------------------------------------------------------------------------------------------------- 
  Line   app/QueryBuilders/ChatLineQueryBuilder.php
 ------ -----------------------------------------------------------------------------------------------------------------------------
  13     Call to private method whereDay() of parent class Illuminate\Database\Eloquent\Builder<Illuminate\Database\Eloquent\Model>.
 ------ -----------------------------------------------------------------------------------------------------------------------------
```

So I read the documentation and went on a google searching spree and ended up with this solution where I had no errors

**ChatLine.php**
```php
<?php

namespace App\Models;

use App\QueryBuilders\ChatLineQueryBuilder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Query\Builder;

/**
 * @method static ChatLineQueryBuilder<ChatLine> query()
 */
class ChatLine extends Model
{
    use HasFactory;

    protected $guarded = [];
    protected $dates = [
        'day'
    ];

    /**
     * @param Builder $query
     * @return ChatLineQueryBuilder
     */
    public function newEloquentBuilder($query): ChatLineQueryBuilder
    {
        return new ChatLineQueryBuilder($query);
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }
}
```

**ChatLineQueryBuilder.php**
```php
<?php

namespace App\QueryBuilders;

use App\Models\User;
use Carbon\CarbonImmutable;
use Illuminate\Database\Eloquent\Builder;

/**
 * @template TModelClass
 * @extends Builder<TModelClass>
 */
class ChatLineQueryBuilder extends Builder
{
    public function forToday(): self
    {
        return $this->whereDay('day', CarbonImmutable::now());
    }

    public function forUser(User $user): self
    {
        return $this->where('user_id', $user->id);
    }
}
```

Great time for level 2 static analysis, but now a lot of new errors occur, they are almost all the same like the one below

```
 ------ ----------------------------------------------------------------------------- 
  Line   app/Http/Controllers/Discord/ChatBucketController.php
 ------ -----------------------------------------------------------------------------
  33     Access to an undefined property Illuminate\Database\Eloquent\Model::$id.
  42     Access to an undefined property Illuminate\Database\Eloquent\Model::$lines.
 ------ -----------------------------------------------------------------------------
```

Here you can see the controller

**ChatBucketController.php**
```php
<?php

namespace App\Http\Controllers\Discord;

use App\Http\Controllers\Controller;
use App\Http\Requests\Discord\DiscordRequest;
use App\Models\ChatLine;
use App\Models\User;
use Carbon\CarbonImmutable;

class ChatBucketController extends Controller
{
    public function __invoke(
        DiscordRequest $request
    ) {
        $chatLines = collect($request->bucket);
        $chatLines = $chatLines->countBy();

        foreach ($chatLines as $discordId => $amount) {
            $user = User::query()->withDiscordId($discordId)->first();

            if (!$user) {
                continue;
            }

            $chatLine = ChatLine::query()
                ->forUser($user)
                ->forToday()
                ->first();

            if (!$chatLine) {
                ChatLine::create([
                    'user_id' => $user->id, //line 33
                    'day' => CarbonImmutable::now(),
                    'lines' => $amount
                ]);

                continue;
            }

            $chatLine->update([
                'lines' => $chatLine->lines + $amount //line 42
            ]);
        }

        return 'ok';
    }
}
```

Addin properties to the models didn't solve the problem. My feeling is that the ->first() method returns a general instance of Model and Larastan isn't aware of the right Model instance. But wathever I try, I cannot seem to solve this issue.

I tried about everything with the  `* @template TModelClass` of the QueryBuilder but no luck.

What am I missing here of what am i doing wrong?

**Comment by szepeviktor** [comment]
Hello Dietger! 👋 

Please see this secret file: https://github.com/nunomaduro/larastan/blob/master/UPGRADE.md#custom-eloquent-builders

**Comment by mdietger** [comment]
I tried those as well but encountered some errors in level 1

**ChatLine.php**
```php
<?php

namespace App\Models;

use App\QueryBuilders\ChatLineQueryBuilder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Query\Builder;

/**
 * @method static ChatLineQueryBuilder<ChatLine> query()
 */
class ChatLine extends Model
{
    use HasFactory;

    protected $guarded = [];
    protected $dates = [
        'day'
    ];

    /**
     * @param Builder $query
     * @return ChatLineQueryBuilder<ChatLine>
     */
    public function newEloquentBuilder($query): ChatLineQueryBuilder
    {
        return new ChatLineQueryBuilder($query);
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }
}
```

**ChatLineQueryBuilder.php**
```php
<?php

namespace App\QueryBuilders;

use App\Models\User;
use Carbon\CarbonImmutable;
use Illuminate\Database\Eloquent\Builder;

/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 * @extends Builder<TModelClass>
 */
class ChatLineQueryBuilder extends Builder
{
    public function forToday(): self
    {
        return $this->whereDay('day', CarbonImmutable::now()); //line 17
    }

    public function forUser(User $user): self
    {
        return $this->where('user_id', $user->id);
    }
}
```

**Larastan output (level 1)**
```
 ------ ------------------------------------------------------------------------------------------------------------------------------------ 
  Line   app/QueryBuilders/ChatLineQueryBuilder.php
 ------ ------------------------------------------------------------------------------------------------------------------------------------
  17     Call to an undefined method App\QueryBuilders\ChatLineQueryBuilder<TModelClass of Illuminate\Database\Eloquent\Model>::whereDay().
 ------ ------------------------------------------------------------------------------------------------------------------------------------
```

Upon digging deeper into this if found that where() is making use of _Illuminate\Database\ **Eloquent** \Builder_ while whereDay() is using _Illuminate\Database\ **Query** \Builder_. Is there a way to use them both?

And even when I continue on level 2 I still get the same result

**Larastan output (level 2)**
```
 ------ -----------------------------------------------------------------------------
  Line   app/Http/Controllers/Discord/ChatBucketController.php
 ------ -----------------------------------------------------------------------------
  33     Access to an undefined property Illuminate\Database\Eloquent\Model::$id.
  42     Access to an undefined property Illuminate\Database\Eloquent\Model::$lines.
```

**Comment by szepeviktor** [comment]
Have you tried removing `@method` from ChatLine?

**Comment by szepeviktor** [comment]
... it could be that `whereDay` cannot be used in a custom builder.

**Comment by szepeviktor** [comment]
Try simplifying your code step-by-step and run static analysis.

```php
// $user = User::query()->withDiscordId($discordId)->first();
$user = User::query()->first();
```

**Comment by mdietger** [comment]
All'right, so I went back and started from scratch. I made a new directory and created a new laravel project and added Larastan.

Now I have this

**ChatLine.php**
```php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

/**
 * @property string $test
 */
class ChatLine extends Model
{
    use HasFactory;
}
```

ChatLineController.php
```php
<?php

namespace App\Http\Controllers;

use App\Models\ChatLine;

class ChatLineController extends Controller
{
    public function index()
    {
        $chatline = ChatLine::first();
        $chatline->test = 'something';
    }
}
```

Great success!

```
 20/20 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
                                                                                                                      
 [OK] No errors

```

Owkay now on to the next step, adding a builder. We do it as documented in the chanelog

**ChatLine.php**
```php
<?php

namespace App\Models;

use App\Builders\ChatLineQueryBuilder;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

/**
 * @property string $test
 */
class ChatLine extends Model
{
    use HasFactory;

    /**
     * @param \Illuminate\Database\Query\Builder $query
     *
     * @return ChatLineQueryBuilder<ChatLine>
     */
    public function newEloquentBuilder($query): ChatLineQueryBuilder
    {
        return new ChatLineQueryBuilder($query);
    }
}
```

**ChatLineQueryBuilder.php**
```php
<?php

namespace App\Builders;

use Illuminate\Database\Eloquent\Builder;

/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 * @extends Builder<TModelClass>
 */
class ChatLineQueryBuilder extends Builder
{

}
```

Now we get the following result

```

 21/21 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ----------------------------------------------------------------------------------------------------------------------------------- 
  Line   Models/ChatLine.php                                                                                                                
 ------ -----------------------------------------------------------------------------------------------------------------------------------
  23     Method App\Models\ChatLine::newEloquentBuilder() should return App\Builders\ChatLineQueryBuilder<App\Models\ChatLine> but returns
         App\Builders\ChatLineQueryBuilder<Illuminate\Database\Eloquent\Model>.
 ------ -----------------------------------------------------------------------------------------------------------------------------------

 [ERROR] Found 1 error
```

So i suppose I need to change?

```
/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 * @extends Builder<TModelClass>
 */
```

to

```
/**
 * @template TModelClass of \App\Models\ChatLine
 * @extends Builder<TModelClass>
 */
```

Is that the correct thing to do?

**Comment by szepeviktor** [comment]
`TModelClass of \Illuminate\Database\Eloquent\Model` is the correct one, it is called PHP generics.
The actual value is determined by the extending class: `ChatLineQueryBuilder<ChatLine>`

**Comment by nagmat84** [comment]
Has this problem been solved? I am experiencing the very same problem but, I even went one step further using with a trait. The idea was to write a custom `FixedQueryBuilder` which extends `Eloquent\Builder` and throws more consistent exceptions in case of an error.

### Example Code

#### The Trait for the Fixed Query Builder

This traits wraps many methods of `Eloquent\Builder`. The illustrate `where` is only an example, there are a lot more following the same pattern.

```php
/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 */
trait FixedQueryBuilderTrait {
  /**
   * Add a basic where clause to the query.
   *
   * @param \Closure|string|array<string>|Expression $column
   * @param mixed                                    $operator
   * @param mixed                                    $value
   * @param string                                   $boolean
   *
   * @return $this
   *
   * @throws QueryBuilderException
   */
  public function where($column, $operator = null, $value = null, $boolean = 'and'): static {
    try {
      // PhpStan reports issue that `parent::where` is a call to an undefined, static method
      return parent::where($column, $operator, $value, $boolean);
    } catch (\Throwable $e) {
      throw new QueryBuilderException($e);
    }
  }
}
```

#### The Fixed Query Builder

This trivial class extends `Eloquent\Builder` and simply uses the trait above. If you wonder why I am using a trait and not defining the methods directly in this class: I need the same trait for another query builder.

```php
/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 * @extends Builder<TModelClass>
 */
class FixedQueryBuilder extends Builder {
  /** @phpstan-use FixedQueryBuilderTrait<TModelClass> */
  use FixedQueryBuilderTrait;
}
```

#### The Trait for Using the Fixed Query Builder

```php
/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 *
 * @method static FixedQueryBuilder<TModelClass> query()                       Begin querying the model.
 * @method static FixedQueryBuilder<TModelClass> with(array|string $relations) Begin querying the model with eager loading.
 */
trait UseFixedQueryBuilder {
  /**
   * @param $query
   * @return FixedQueryBuilder<TModelClass>
   */
  public function newEloquentBuilder($query): FixedQueryBuilder {
    // PhpStan reports an issue that the actual return value `FixedQueryBuilder` does not match `FixedQueryBuilder<TModelClass>`
    return new FixedQueryBuilder($query);
  }
}
```

#### A Model Class

```php
class MyModel extends Model {
  /** @phpstan-use UseFixedQueryBuilder<MyModel> */
  use UseFixedQueryBuilder;
}
```

### The Problems

There is bunch of problems involved here

 1. In `FixedQueryBuilderTrait` all the calls to the parent methods are reported as undefined. I can somehow live with that. As it is unclear when and how the trait might be used, `$parent` is indeed ambiguous.
 2. In `UseFixedQueryBuilder::newEloquentBuilder` PhpStan reports an issue that `new FixedQueryBuilder` is of incorrect type. I can ignore that, too, if I have to.
 3. However most of the issue are reported for `MyModel` or more precisely when I try to use it.
     1. Any query method on a query for `MyModel` which has not been overwritten in `FixedQueryBuilderTrait` is reported as undefined even though `FixedQueryBuilder` extends `Eloquent\Builder`. For example,
        ```php
        MyModel::query()->where('foo', '=', 42)->orderBy('bar')->get();
        ```
        reports `orderBy` as undefined. This problem can be seemingly fixed if `FixedQueryBuilder` additionally declares `Laravel\Query\Builder` as a mixin, but these creates other problems (see below). Moreover, this should not be necessary, because `FixedQueryBuilder` inherits `Eloquent\Builder`.
     2. The template parameter `TModelClass` is not substituted by the correct model. For example consider the following code
        ```php
        class SomeClass {
          /** @var Collection<MyModel> */
          protected Collection $myModels;
          
          public function doSometing(): void {
            $this->myModels = MyModel::query()->get();
          }
        }
        ```
        PhpStan reports that `SomeClass::myModels` is of type `Collection<MyModel>` but `Collection<TModelClass>` has been given. In this context it might be worth noting that my IDE does not like the following
        code
        ```php
        /**
         * @template TModelClass of \Illuminate\Database\Eloquent\Model
         *
         * @method static FixedQueryBuilder<TModelClass> query()                       Begin querying the model.
         * @method static FixedQueryBuilder<TModelClass> with(array|string $relations) Begin querying the model with eager loading.
         */
        trait UseFixedQueryBuilder { ... }
        ```
        My IDE complains that `TModelClass` in line 5+6 (i.e. in the `@method` annotations) would be undefined. It seems as if `TModelClass` is used a class name, not as a template parameter.

It is not a work-around to mixin `Laravel\Query\Builder` into `FixedQueryBuilder` like this
```php
/**
 * @template TModelClass of \Illuminate\Database\Eloquent\Model
 * @extends Builder<TModelClass>
 * @mixin Query\Builder
 */
class FixedQueryBuilder extends Eloquent\Builder {
  /** @phpstan-use FixedQueryBuilderTrait<TModelClass> */
  use FixedQueryBuilderTrait;
}
```
While it removes the error message about undefined methods, it creates hundreds of new messages due to wrong return types. Note, most query methods return `$this` such that the methods can be changed. If I mixin `Query\Builder`, then `$this` is resolved to `Query\Builder` and not to `FixedQueryBuilder` anymore. In the example above,
```php
MyModel::query()->where('foo', '=', 42)->orderBy('bar')->get();
```
the error message about `orderBy` being undefined vanishes, but instead `get()` is assumed to return an `array` and not `Collection<MyModel>`, because the result of `orderBy` is wrongly assumed to be `Query\Builder`.
Anyway, `@mixin` should not be required, because we properly inherit from `Eloquent\Builder`.

**Reply by canvural** [reply]
This is kind of hard to keep track for me. Can you please create a new repo that demonstrates these issue? Thanks.

**Comment by szepeviktor** [comment]
Do all problems disappear when you copy your trait's code to all classes using that trait?

**Reply by nagmat84** [reply]
Let me try that, but that may take a while, because the real project is huge and I need to create a dummy project first. That may take a while.

**Reply by nagmat84** [reply]
I gave it a quick try. While some errors disappears, some do not and others pop up. Moreover, I discovered at least one definite bug in Larastan, see here https://github.com/nunomaduro/larastan/discussions/1275#discussioncomment-2933818.

I will create a repository with a stripped down version of our developing repository as @canvural asked for.

**Reply by nagmat84** [reply]
I have promised a repository which illustrates the problem. I haven't yet created one, because I would like to wait until https://github.com/nunomaduro/larastan/pull/1285 will have been merged. I believe that quite a share of the errors messages can be traced back to the `Relation`-related classes which do not accurately distinguish between the type of the related class and the type of the declaring class.

This does not mean that I expect all problems to be solved by https://github.com/nunomaduro/larastan/pull/1285. Possibly, some errors are indeed caused by the trait instead of proper inheritance. But I want to get rid of the obvious bugs first and then see what errors still remains.

**Reply by nagmat84** [reply]
As requested by @canvural [in this post](https://github.com/nunomaduro/larastan/discussions/1239#discussioncomment-2933276) I created an example repository here: https://github.com/nagmat84/larastan-example

Steps to reproduce the errors:

 1. Checkout the repository
 2. Install all dependencies via `composer install`
 3. Run PHPStan via `composer phpstan`

You get the following results

```
 11/11 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ---------------------------------------------------------------------------------------------------------- 
  Line   Http/Controllers/Controller.php                                                                           
 ------ ---------------------------------------------------------------------------------------------------------- 
  18     Call to private method whereBetween() of parent class Illuminate\Database\Eloquent\Builder<TNodeModel of  
         App\Models\Contracts\Node&Illuminate\Database\Eloquent\Model>.                                            
 ------ ---------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------ 
  Line   Models/Concerns/NodeTrait.php (in context of class App\Models\Node)                                               
 ------ ------------------------------------------------------------------------------------------------------------------ 
  53     Method App\Models\Node::newEloquentBuilder() should return App\Database\NodeBuilder<App\Models\Node> but returns  
         App\Database\NodeBuilder<App\Models\Contracts\Node&Illuminate\Database\Eloquent\Model>.                           
  62     Method App\Models\Node::newCollection() should return App\Database\NodeCollection<App\Models\Node> but returns    
         App\Database\NodeCollection<App\Models\Contracts\Node&Illuminate\Database\Eloquent\Model>.                        
  72     Method App\Models\Node::siblings() should return App\Database\NodeBuilder<App\Models\Node> but returns            
         Illuminate\Database\Eloquent\Builder|static(App\Models\Node).                                                     
 ------ ------------------------------------------------------------------------------------------------------------------ 
                                                                                                                        
 [ERROR] Found 4 errors
```

The first issue report about the private method is definitively a bug. I also reported it in the issue tracker as https://github.com/nunomaduro/larastan/issues/1289. I am not so sure whether the last three issue reports also indicate a bug or if I made an annotation error. I suspect the former.

_Note 1:_ This has nothing to do with PHPStan strict rules. As you can see in `phpstan.neon` in the repository, strict rules are not used.

_Note 2:_ This is only the first batch of problems. PHPStan emits more (and similar) issue reports as soon as there are custom relations involved. However as a starter, I wanted to keep the example simple.
